File: dotnet/src/Envdoctor/Core/ModelAssembler.cs
using Envdoctor.Config;
using Envdoctor.Models;
using Envdoctor.Parsers;
using Envdoctor.Utils;

namespace Envdoctor.Core;

/// Assemble a `ProjectModel` from discovered file paths.
public static class ModelAssembler
{
    public static ProjectModel AssembleModel(string rootDir, EnvdoctorConfig config, IReadOnlyList<string> discovered)
    {
        var registry = ParserRegistry.DefaultRegistry(config.SourceExtensions);
        var envFiles = new List<EnvironmentFile>();
        var composeFiles = new List<EnvironmentFile>();
        var actionFiles = new List<EnvironmentFile>();
        var k8sFiles = new List<EnvironmentFile>();
        var sourceFiles = new List<EnvironmentFile>();
        var allFiles = new List<EnvironmentFile>();
        var parseErrors = new List<ParseError>();

        foreach (var path in discovered)
        {
            string content;
            try
            {
                content = File.ReadAllText(path);
            }
            catch (Exception e)
            {
                parseErrors.Add(new ParseError { FilePath = path, Error = $"Cannot read file: {e.Message}" });
                continue;
            }

            // Skip if ignored by config
            if (config.IgnoreFiles.Count > 0)
            {
                var rel = Discover.RelativeTo(rootDir, path);
                if (rel is not null && Glob.MatchesAnyGlob(config.IgnoreFiles, rel))
                    continue;
            }

            EnvironmentFile? parsed = null;
            foreach (var parser in registry)
            {
                if (parser.MatchPath(path))
                {
                    parsed = parser.Parse(content, path);
                    break;
                }
            }

            if (parsed is not null)
            {
                switch (parsed.Format)
                {
                    case FileFormat.Dotenv: envFiles.Add(parsed); break;
                    case FileFormat.DockerCompose: composeFiles.Add(parsed); break;
                    case FileFormat.GithubActions: actionFiles.Add(parsed); break;
                    case FileFormat.Kubernetes: k8sFiles.Add(parsed); break;
                    case FileFormat.Source: sourceFiles.Add(parsed); break;
                }
                allFiles.Add(parsed);
            }
        }

        // Apply ignoreVariables to parsed variables
        envFiles = ApplyIgnoreVariables(envFiles, config);
        composeFiles = ApplyIgnoreVariables(composeFiles, config);
        actionFiles = ApplyIgnoreVariables(actionFiles, config);
        k8sFiles = ApplyIgnoreVariables(k8sFiles, config);
        sourceFiles = ApplyIgnoreVariables(sourceFiles, config);

        // Apply environment overrides if configured
        envFiles = ApplyEnvironmentOverrides(envFiles, config);

        return new ProjectModel
        {
            RootDir = rootDir,
            Config = config,
            EnvFiles = envFiles,
            ComposeFiles = composeFiles,
            ActionFiles = actionFiles,
            K8sFiles = k8sFiles,
            SourceFiles = sourceFiles,
            AllFiles = allFiles,
            ParseErrors = parseErrors,
        };
    }

    /// Remove variables whose names match `ignoreVariables` globs.
    private static List<EnvironmentFile> ApplyIgnoreVariables(List<EnvironmentFile> files, EnvdoctorConfig config)
    {
        if (config.IgnoreVariables.Count == 0)
            return files;
        foreach (var f in files)
        {
            f.Variables.RemoveAll(v => Glob.MatchesAnyGlob(config.IgnoreVariables, v.Name));
            f.Usages.RemoveAll(v => Glob.MatchesAnyGlob(config.IgnoreVariables, v.Name));
        }
        return files;
    }

    /// If `environments` is configured, override the environment label for
    /// files matching the given globs.
    private static List<EnvironmentFile> ApplyEnvironmentOverrides(List<EnvironmentFile> files, EnvdoctorConfig config)
    {
        if (config.Environments is null)
            return files;
        foreach (var f in files)
        {
            var rel = f.FilePath;
            foreach (var (envLabel, globs) in config.Environments)
            {
                if (Glob.MatchesAnyGlob(globs, rel))
                {
                    f.Environment = envLabel;
                    break;
                }
            }
        }
        return files;
    }
}

File: dart/lib/src/core/model_assembler.dart
/// Assemble a [ProjectModel] from discovered file paths.
library;

import 'dart:io';

import '../config/config.dart';
import '../models/environment_file.dart';
import '../models/project_model.dart';
import '../parsers/parser.dart';
import '../utils/glob.dart';
import '../utils/paths.dart';

ProjectModel assembleModel(
    String rootDir, EnvdoctorConfig config, List<String> discovered) {
  final registry = defaultRegistry(config.sourceExtensions);
  final envFiles = <EnvironmentFile>[];
  final composeFiles = <EnvironmentFile>[];
  final actionFiles = <EnvironmentFile>[];
  final k8sFiles = <EnvironmentFile>[];
  final sourceFiles = <EnvironmentFile>[];
  final allFiles = <EnvironmentFile>[];
  final parseErrors = <ParseError>[];

  for (final path in discovered) {
    final String content;
    try {
      content = File(path).readAsStringSync();
    } catch (e) {
      parseErrors.add(ParseError(path, 'Cannot read file: $e'));
      continue;
    }

    // Skip if ignored by config.
    if (config.ignoreFiles.isNotEmpty) {
      final rel = relativeTo(rootDir, path);
      if (rel != null && matchesAnyGlob(config.ignoreFiles, rel)) continue;
    }

    EnvironmentFile? parsed;
    for (final parser in registry) {
      if (parser.matchPath(path)) {
        parsed = parser.parse(content, path);
        break;
      }
    }

    if (parsed != null) {
      switch (parsed.format) {
        case FileFormat.dotenv:
          envFiles.add(parsed);
        case FileFormat.dockerCompose:
          composeFiles.add(parsed);
        case FileFormat.githubActions:
          actionFiles.add(parsed);
        case FileFormat.kubernetes:
          k8sFiles.add(parsed);
        case FileFormat.source:
          sourceFiles.add(parsed);
      }
      allFiles.add(parsed);
    }
  }

  // Apply ignoreVariables to parsed variables.
  _applyIgnoreVariables(envFiles, config);
  _applyIgnoreVariables(composeFiles, config);
  _applyIgnoreVariables(actionFiles, config);
  _applyIgnoreVariables(k8sFiles, config);
  _applyIgnoreVariables(sourceFiles, config);

  // Apply environment overrides if configured.
  _applyEnvironmentOverrides(envFiles, rootDir, config);

  final model = ProjectModel(rootDir: rootDir, config: config)
    ..envFiles = envFiles
    ..composeFiles = composeFiles
    ..actionFiles = actionFiles
    ..k8sFiles = k8sFiles
    ..sourceFiles = sourceFiles
    ..allFiles = allFiles
    ..parseErrors = parseErrors;
  return model;
}

/// Remove variables whose names match `ignoreVariables` globs.
void _applyIgnoreVariables(List<EnvironmentFile> files, EnvdoctorConfig config) {
  if (config.ignoreVariables.isEmpty) return;
  for (final f in files) {
    f.variables
        .removeWhere((v) => matchesAnyGlob(config.ignoreVariables, v.name));
    f.usages.removeWhere((v) => matchesAnyGlob(config.ignoreVariables, v.name));
  }
}

/// If `environments` is configured, override the environment label for
/// files matching the given globs.
void _applyEnvironmentOverrides(
    List<EnvironmentFile> files, String rootDir, EnvdoctorConfig config) {
  final environments = config.environments;
  if (environments == null) return;
  for (final f in files) {
    final rel = relativeTo(rootDir, f.filePath) ?? f.filePath;
    for (final entry in environments.entries) {
      if (matchesAnyGlob(entry.value, rel)) {
        f.environment = entry.key;
        for (final v in f.variables) {
          for (final o in v.origins) {
            o.environment = entry.key;
          }
        }
        for (final v in f.usages) {
          for (final o in v.origins) {
            o.environment = entry.key;
          }
        }
        break;
      }
    }
  }
}
